### PR TITLE
[stable/nginx-ingress] Added `extraVolumes` configuration point to attach additional

### LINKED ISSUE #4417: [stable/nginx-ingress] Adding additional volume mounts for ModSecurity configuration
**Is this a request for help?**: Yes

---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one): BUG REPORT

**Version of Helm and Kubernetes**: 
Client: &version.Version{SemVer:"v2.8.1", GitCommit:"6af75a8fd72e2aa18a2b278cfe5c7a1c5feca7f2", GitTreeState:"clean"}
Server: &version.Version{SemVer:"v2.8.1", GitCommit:"6af75a8fd72e2aa18a2b278cfe5c7a1c5feca7f2", GitTreeState:"clean"}

**Which chart**:
stable/nginx-ingress

**What happened**:
I would like to have additional volumes mounted to Nginx for including custom configuration files for ModSecurity rules. Prior to using charts, our deployment included ConfigMaps mounted as volumes under `/etc/nginx/` and `/etc/nginx/modsecurity`like so:

```
volumeMounts:
          - name: modsecurity-config
            mountPath: /etc/nginx/modsecurity/modsecurity.conf
            subPath: modsecurity.conf
          - name: crs-setup-config
            mountPath: /etc/nginx/owasp-modsecurity-crs/crs-setup.conf
            subPath: crs-setup.conf
          - name: crs-exceptions-rules-before-config
            mountPath: /etc/nginx/owasp-modsecurity-crs/rules/REQUEST-900-EXCLUSION-RULES-BEFORE-CRS.conf
            subPath: REQUEST-900-EXCLUSION-RULES-BEFORE-CRS.conf
          - name: crs-exceptions-rules-after-config
            mountPath: /etc/nginx/owasp-modsecurity-crs/rules/RESPONSE-999-EXCLUSION-RULES-AFTER-CRS.conf
            subPath: RESPONSE-999-EXCLUSION-RULES-AFTER-CRS.conf
```
However, the Helm chart does not provide an option to mount additional configuration as volumes.

**What you expected to happen**:
Allow additional configuration (configmaps) to be mounted as volumes.

**How to reproduce it** (as minimally and precisely as possible):
Try mounting a configmap as a file under `/etc/nginx/`

**Anything else we need to know**:

### PR DESCRIPTION
**Pull request by logicfox**
volumes to the nginx-controller pod in order to address
use-cases where additional configuration files may be needed,
as in the case of custom ModSecurity configuration.

**What this PR does / why we need it**: 
Added `extraVolumes` configuration point to attach additional volumes to the nginx-controller pod in order to address use-cases where additional configuration files may be needed, as in the case of custom ModSecurity configuration.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #4417

**Special notes for your reviewer**:
The specification of additional volumes is now generic and supersedes the current mechanism of mounting the nginx.tmpl. The current way of specifying this file has been preserved to maintain compatibility with existing deployments.

### CODE REVIEW

File: stable/nginx-ingress/Chart.yaml [hunk 1]
```diff
@@ -1,5 +1,5 @@
 name: nginx-ingress
-version: 0.11.2
+version: 0.11.3
```
Comment (timstoop) [inline stable/nginx-ingress/Chart.yaml]: This PR adds a new feature, so you can bump the minor version.

**Review by timstoop — CHANGES_REQUESTED** [review]
So I don't personally use this controller that much, but is there a specific reason why you wouldn't want to add those ModSecurity files via a ConfigMap? I don't think it's very user friendly to require users to create their own Volume with the correct data in there, when they simply want to add some files. I would prefer to see this rewritten to make use of ConfigMaps instead, as it's much easier to add those.

**Comment by logicfox** [comment]
The ModSecurity WAF ships with the default OWASP ruleset which unfortunately is quite strict resulting in false positives. The typical solution suggested is to customise the ruleset. 

This often results in modifying/overriding multiple files in the OWASP ruleset located in `/etc/nginx/owasp-modsecurity-crs`. Putting this kind of configuration in the `values.yaml` would bloat it considerably. 

Besides, a further resolving was that the addition of custom volumes and mounts would allow for edge use-cases driven by configuration, like the WAF, to work without additional changes to the chart. 

The typical ruleset that ships with the ingress controller is here:
https://github.com/SpiderLabs/owasp-modsecurity-crs/tree/v3.0/master/rules

**Comment by timstoop** [comment]
I understand, but I wonder if for the regular user adding a Volume with those files is really that easy. You can create a ConfigMap using the source files instead and mount that one. I'd expect that to be a lot easier to do for most people.

**Comment by unguiculus** [comment]
Closing this in favor of #3983.

@timstoop You may want to follow the discussion here: #4283.

**Pull request closed without merging**